recipes/sqlite3/all/CMakeLists.txt

```diff
@@ -94,9 +110,8 @@ install(DIRECTORY source_subfolder/ DESTINATION include
         FILES_MATCHING PATTERN "*.h")
 
 if(SQLITE3_BUILD_EXECUTABLE)
-    add_executable(sqlite3-bin source_subfolder/shell.c source_subfolder/sqlite3.c)
-    target_compile_definitions(sqlite3-bin PRIVATE "SQLITE_THREADSAFE=0")
-    target_link_libraries(sqlite3-bin PRIVATE ${CMAKE_DL_LIBS})
+    add_executable(sqlite3-bin source_subfolder/shell.c)
+    target_link_libraries(sqlite3-bin PRIVATE ${PROJECT_NAME})
```

Comment:
These files are never used by Conan when consuming this recipe. Conan uses only the actual binary and the information provided in `cpp_info`.

Reply:
But if shared, executable depends on shared sqlite3. So `RPATH` must be `$ORIGIN/../lib` otherwise executable will fail at runtime if used as a build requirement. It's just less pain to always build this executable from sources without linking to sqlite3 lib.

Reply:
Conan will add the path to the `bin` folder (`PATH`, `DYLD_LIBRARY_PATH`) and there the shared library will be found. Using two profiles, these environment variables are populated with the information coming from the build-requires, using one profile you need the `run_environment=True`, or using a `virtualenv` generator if running locally... but in all cases the information is available and ready to use.

Reply:
Coul you test with `proj` recipe? It looks for `sqlite3` executable inside CMake, pretty sure it breaks if sqlite3 is shared with this PR.

Reply:
Indeed, proj does not build with the following `conan create recipes/proj/7.x.x proj/7.1.0@ -o sqlite3:shared=True`, however, given that I explicitly set that sqlite3 should be shared, I strongly believe that sqlite3 should be used as a shared library. In which case, either the proj recipe (and the like) should use conan's run environment to insure `LD_LIBRARY_PATH` is set, or sqlite should have it's rpath set. I'd be inclined towards the former, given that I know of no examples of conanfiles that set the rpath.

Reply:
> Is it really an improvement?
It most certainly is. Not only does it actually match the intent of building sqlite as a shared library, but it also insures that the actual options that were specified (e.g. `enable_fts5`) apply to the executable as well and not just the library.

Reply:
Yes, calling executable from inside CMake is a problem if they depend on a shared library. 

**MacOS**: even if we populate the `DYLD_LIBRARY_PATH`, due to SIP protection it is erased and not available inside CMake, so it cannot be propagated to the environment when calling Sqlite3 and it will fail. The only solution I came up with is to use _wrappers_ around executables (https://github.com/conan-io/conan/pull/7324).

**Linux**: I would really expect it to work if the environment is populated before running CMake, does it fail?

In this PR you can read about the same issue for `protoc` (https://github.com/conan-io/conan-center-index/pull/2488).

**rpath**: @ohanar, we cannot use it, Conan distributes binaries compiled in different machine, the path won't match.

Reply:
> **rpath**: @ohanar, we cannot use it, Conan distributes binaries compiled in different machine, the path won't match.

In this case, because the sqlite executable only depends upon the sqlite library and system libraries, you can get away with a relative rpath. So on linux `$ORIGIN/../lib` and on MacOS `@executable_loader/../lib`.

Reply:
> **Linux**: I would really expect it to work if the environment is populated before running CMake, does it fail?
> 
Doing it manually works fine:
```sh
$ conan install recipes/proj/7.x.x proj/7.1.0@ -o sqlite3:shared=True -g virtualbuildenv
$ . ./activate_build.sh
$ conan create recipes/proj/7.x.x proj/7.1.0@ -o sqlite3:shared=True
```

Reply:
Wait, it means that all the recipes relying of a build requirement where binaries can be linked to their shared libs, should be updated to add `virtualbuildenv` and then calling `activate_build.sh` at some point (first step in build method?)?

Reply:
If used inside Conan they will work (except Macos due to SIP protection), if you use them in a local build you need to populate the environment (and you have the same Macos-SIP problem).

Reply:
So all recipes where these binaries are called inside CMake (let's inside their own build system, even if scripting of this process is managed by conan)?

Reply:
Yes. The problem is that you define your `DYLD_LIBRARY_PATH` with the path to your libraries, then you run CMake using those libraries, but inside CMake your values are no longer available... MacOS uses this protection so you cannot _inject_ your malicious libraries to other programs called by the one you are actually calling. 🤷 

The only workaround I can imagine so far is to use `wrappers` around executables (CMake will run the wrapper instead of the actual executable, and the wrapper will handle the environment). It has also another big advantage from the POV of Conan which is that you are not adding all together the environment from different build-requires.

Reply:
I've managed to build proj with sqlite3 shared. In proj recipe, I've added `with tools.environment_append(RunEnvironment(self).vars):` in build method.

```
with tools.environment_append(RunEnvironment(self).vars):
    cmake = self._configure_cmake()
    cmake.build()
```

I don't know if there is any side effects in some situations.